core/unit_test/tstBufferedData.hpp

```diff
@@ -0,0 +1,196 @@
+                    EXPECT_EQ( slice_0( idx, i, j, k ), fval * ( i + j + k ) );
+
+        // Member 1.
+        EXPECT_EQ( slice_1( idx ), ival );
+
+        // Member 2.
+        for ( int i = 0; i < dim_1; ++i )
+            EXPECT_EQ( slice_2( idx, i ), dval * i );
+
+        // Member 3.
+        for ( int i = 0; i < dim_1; ++i )
+            for ( int j = 0; j < dim_2; ++j )
+                EXPECT_EQ( slice_3( idx, i, j ), dval * ( i + j ) );
+    }
+}
+
+void testBufferedDataCreation()
+{
+    // Create an AoSoA
+    const int vector_length = 16;
+
+    // Data dimensions.
+    const int dim_1 = 3;
+    const int dim_2 = 2;
+    const int dim_3 = 4;
+
+    // Declare data types.
+    using DataTypes = Cabana::MemberTypes<float[dim_1][dim_2][dim_3], int,
+                                          double[dim_1], double[dim_1][dim_2]>;
+
+    // Declare the AoSoA type.
+    using AoSoA_t = Cabana::AoSoA<DataTypes, TEST_MEMSPACE, vector_length>;
+    std::string label = "sample_aosoa";
+    int num_data = 512;
+    AoSoA_t aosoa( label, num_data );
+
+    // Start by only buffering over one AoSoA at a time for stress test
+    const int max_buffered_tuples = 8 * vector_length;
+
+    // emulate a minimum of triple buffering?
+    const int buffer_count = 3;
+
+    // Hard code into OpenMP space for now
+    // TODO: specify the exec space via a test param
+    using target_exec_space = TEST_EXECSPACE;
+
+    // Init the AoSoA data
+    auto mirror =
+        Cabana::create_mirror_view_and_copy( Kokkos::HostSpace(), aosoa );
+    auto mirror_slice_0 = Cabana::slice<0>( mirror );
+    auto mirror_slice_1 = Cabana::slice<1>( mirror );
+    auto mirror_slice_2 = Cabana::slice<2>( mirror );
+    auto mirror_slice_3 = Cabana::slice<3>( mirror );
+
+    // Initialize data with the rank accessors.
+    float fval = 3.4;
+    double dval = 1.23;
+    int ival = 1;
+    for ( std::size_t idx = 0; idx != aosoa.size(); ++idx )
+    {
+        // Member 0.
+        for ( int i = 0; i < dim_1; ++i )
+            for ( int j = 0; j < dim_2; ++j )
+                for ( int k = 0; k < dim_3; ++k )
+                    mirror_slice_0( idx, i, j, k ) = fval * ( i + j + k );
+
+        // Member 1.
+        mirror_slice_1( idx ) = ival;
+
+        // Member 2.
+        for ( int i = 0; i < dim_1; ++i )
+            mirror_slice_2( idx, i ) = dval * i;
+
+        // Member 3.
+        for ( int i = 0; i < dim_1; ++i )
+            for ( int j = 0; j < dim_2; ++j )
+                mirror_slice_3( idx, i, j ) = dval * ( i + j );
+    }
+    Cabana::deep_copy( aosoa, mirror );
+    checkDataMembers( aosoa, fval, dval, ival, dim_1, dim_2, dim_3 );
+
+    using buf_t = Cabana::BufferedAoSoA<max_buffered_tuples, buffer_count,
+                                        target_exec_space, AoSoA_t>;
+
+    buf_t buffered_aosoa( aosoa );
+
+    // int num_buffers = buffered_aosoa.get_buffer_count();
+
+    // TODO: are the fences needed before we start?
```

Comment:
Shouldn't be - even if you use streams for example you should be able to control how the streams interact and avoid it